doc/pam_chatty.8
'\" t
.\"     Title: pam_chatty
.\"    Author: [FIXME: author] [see http://www.docbook.org/tdg5/en/html/author]
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 2020-03-05
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "PAM_CHATTY" "8" "2020\-03\-05" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
pam_chatty \- A PAM test module that generates PAM info and error messages\&.
.SH "SYNOPSIS"
.sp
pam_chatty\&.so [\&...]
.SH "DESCRIPTION"
.sp
The \fBpam_chatty\fR module makes it easy to replicate the behaviour of PAM modules that generate a lot of output, in order to ensure that PAM clients react as expected\&.
.SH "IMPORTANT"
.sp
pam_chatty is a \fBtest tool\fR\&. It should be considered completely insecure and never used outside test environments! As you\(cqll see when reading description of the options, many of them don\(cqt make any sense in the real world and were added just to make tests possible\&.
.sp
The pam_chatty module generates a lot of PAM info and error output, when its authentication method is called, and then returns success\&.
.SH "OPTIONS"
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBnum_lines=NUMBER\fR
\- the number of lines of messages to output, for each one of the enabled types of messages selected\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBerror\fR
\- whether to output error level messages\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBinfo\fR
\- whether to output info level messages\&.
.RE
.SH "MODULE TYPES PROVIDED"
.sp
Only the \fBauth\fR module type is supported\&.
.SH "EXAMPLE"
.sp
.if n \{\
.RS 4
.\}
.nf
auth            required        pam_chatty\&.so num_lines=16 info error
.fi
.if n \{\
.RE
.\}


CHANGELOG
ChangeLog
==========

version 1.1.3 (released 2020-03-26)
  * Fixed paths in pkgconfig and cmake config files

version 1.1.2 (released 2020-03-24)
  * Fix manpage installation

version 1.1.1 (released 2020-03-24)
  * Fixed pam_wrapper on Ubuntu
  * Improved PAM detection for openSUSE Tumbleweed

version 1.1.0 (released 2020-03-20)
  * Added support for pam_start_confdir()
  * Added pam_chatty module
  * Added gitlab CI support
  * Fixed crash when a PAM module outputs too much data

version 1.0.7 (released 2018-09-26)
  * Added support for running with AddressSanitizer
  * Added logging for pam_set_items module
  * Fixed building python modules
  * Fixed pam conversation in libpamtest

version 1.0.6 (released 2018-03-27)
  * Improved file copy
  * Fixed build warnings

version 1.0.5 (released 2018-02-22)
  * Added support to build python2 and python3 module at the same time
  * Improved pam test directory creating
  * Fixed python 2.6 compatibilty
  * Fixed some build issues on FreeBSD

version 1.0.4 (released 2017-05-15)
  * Fix build on OpenBSD
  * Fix a resource leak

version 1.0.3 (released 2017-04-06)
  * Fixed some build issues with strict compiler flags
  * Logging to syslog is disabled by default

version 1.0.2 (released 2016-05-24)
  * Fixed pam_wrapper on some BSDs
  * Fixed simple conversation in libpamtest

version 1.0.1 (released 2016-01-18)
  * Fixed issue with audit_open() and sshd
  * Fixed several issues found by Coverity
  * Fixed python linking issues

version 1.0.0 (released 2016-01-14)
  * Initial release
    - pam_wrapper
    - libpamtest
    - pypamtest
    - pam_matrix
    - pam_get_items
    - pam_set_items


src/modules/pam_chatty.c
/*
 * Copyright (c) 2015 Andreas Schneider <asn@samba.org>
 * Copyright (c) 2015 Jakub Hrozek <jakub.hrozek@posteo.se>
 * Copyright (c) 2020 Bastien Nocera <hadess@hadess.net>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"

#include <sys/param.h>
#include <sys/types.h>
#include <sys/stat.h>

#include <pwd.h>
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <unistd.h>
#include <ctype.h>
#include <errno.h>
#include <time.h>
#include <stdint.h>

#ifndef discard_const
#define discard_const(ptr) ((void *)((uintptr_t)(ptr)))
#endif

#ifndef discard_const_p
#define discard_const_p(type, ptr) ((type *)discard_const(ptr))
#endif

#ifdef HAVE_SECURITY_PAM_APPL_H
#include <security/pam_appl.h>
#endif
#ifdef HAVE_SECURITY_PAM_MODULES_H
#include <security/pam_modules.h>
#endif
#ifdef HAVE_SECURITY_PAM_EXT_H
#include <security/pam_ext.h>
#endif

#include "pwrap_compat.h"

#define ERROR_KEY	"error"
#define INFO_KEY	"info"
#define NUM_LINES_KEY	"num_lines="

#define DEFAULT_NUM_LINES 3

/* We only return up to 16 messages from the PAM conversation.
 * Value from src/python/pypamtest.c */
#define PAM_CONV_MSG_MAX        16

#define PAM_CHATTY_FLG_ERROR	(1 << 0)
#define PAM_CHATTY_FLG_INFO	(1 << 1)

static int pam_chatty_conv(pam_handle_t *pamh,
			   const int msg_style,
			   const char *msg)
{
	int ret;
	const struct pam_conv *conv = NULL;
	const struct pam_message *mesg[1];
	struct pam_response *r = NULL;
	struct pam_message *pam_msg = NULL;

	ret = pam_get_item(pamh, PAM_CONV, (const void **) &conv);
	if (ret != PAM_SUCCESS) {
		return ret;
	}

	pam_msg = calloc(sizeof(struct pam_message), 1);
	if (pam_msg == NULL) {
		return PAM_BUF_ERR;
	}

	pam_msg->msg_style = msg_style;
	pam_msg->msg = discard_const_p(char, msg);

	mesg[0] = (const struct pam_message *) pam_msg;
	ret = conv->conv(1, mesg, &r, conv->appdata_ptr);
	free(pam_msg);

	return ret;
}

/* Evaluate command line arguments and store info about them in the
 * pam_matrix context
 */
static uint32_t parse_args(int argc,
			   const char *argv[],
			   unsigned int *num_lines)
{
	uint32_t flags = 0;

	*num_lines = DEFAULT_NUM_LINES;

	for (; argc-- > 0; ++argv) {
		if (strncmp(*argv, NUM_LINES_KEY, strlen(NUM_LINES_KEY)) == 0) {
			if (*(*argv+strlen(NUM_LINES_KEY)) != '\0') {
				*num_lines = atoi(*argv+strlen(NUM_LINES_KEY));
				if (*num_lines <= DEFAULT_NUM_LINES)
					*num_lines = DEFAULT_NUM_LINES;
				if (*num_lines > PAM_CONV_MSG_MAX)
					*num_lines = PAM_CONV_MSG_MAX;
			}
		} else if (strncmp(*argv, ERROR_KEY,
				   strlen(ERROR_KEY)) == 0) {
			flags |= PAM_CHATTY_FLG_ERROR;
		} else if (strncmp(*argv, INFO_KEY,
				   strlen(INFO_KEY)) == 0) {
			flags |= PAM_CHATTY_FLG_INFO;
		}
	}

	return flags;
}

PAM_EXTERN int
pam_sm_authenticate(pam_handle_t *pamh, int flags,
		    int argc, const char *argv[])
{
	uint32_t optflags;
	unsigned int num_lines;

	(void) flags;	/* unused */

	optflags = parse_args (argc, argv, &num_lines);
	if (optflags & PAM_CHATTY_FLG_INFO) {
		unsigned int i;

		for (i = 0; i < num_lines; i++) {
			pam_chatty_conv(pamh,
					PAM_TEXT_INFO,
					"Authentication succeeded");
		}
	}

	if (optflags & PAM_CHATTY_FLG_ERROR) {
		unsigned int i;

		for (i = 0; i < num_lines; i++) {
			pam_chatty_conv(pamh,
					PAM_ERROR_MSG,
					"Authentication generated an error");
		}
	}

	return PAM_SUCCESS;
}
